vm/src/vm.rs

```diff
@@ -224,74 +224,150 @@ impl<ENV> GVm<ENV> {
+                    let (car2, cdr2) = val2.get_pair(self).expect("Must be a pair or list!");
+                    val = self.is_equal_pair(car1, car2)?;
+                    if val == Value::True {
+                        val = self.is_equal_pair(cdr1, cdr2)?;
+                    }
+                }
+                (Value::Map(m1), Value::Map(m2)) => {
+                    let m1 = self.heap().get_map(m1);
+                    let m2 = self.heap().get_map(m2);
+                    if m1.len() == m2.len() {
+                        if m1.is_empty() {
+                            val = Value::True;
+                        } else {
+                            // must set val to false in two instances because
+                            // its possible a previous iteration set val to true.
+                            for (k, v) in m1.iter() {
+                                if let Some(v2) = m2.get(k) {
+                                    if self.is_equal_pair(*v, *v2)? == Value::False {
                                         val = Value::False;
                                         break;
+                                    } else {
+                                        val = Value::True;
                                     }
+                                } else {
+                                    val = Value::False;
+                                    break;
                                 }
                             }
                         }
                     }
                 }
-                Value::Pair(_) | Value::List(_, _) => {
-                    // XXX use iterators to reduce recursion?
-                    // Make sure pair iter will work for non-lists...
-                    if matches!(val2, Value::Pair(_) | Value::List(_, _)) {
-                        let (car1, cdr1) = val1.get_pair(self).expect("Must be a pair or list!");
-                        let (car2, cdr2) = val2.get_pair(self).expect("Must be a pair or list!");
-                        val = self.is_equal_pair(car1, car2)?;
-                        if val == Value::True {
-                            val = self.is_equal_pair(cdr1, cdr2)?;
-                        }
-                    }
+                (Value::Value(h1), Value::Value(h2)) => {
+                    let v1 = self.get_value(h1);
+                    let v2 = self.get_value(h2);
+                    val = self.is_equal_pair(v1, v2)?;
+                }
+                (val1, Value::Value(h2)) => {
+                    let v2 = self.get_value(h2);
+                    val = self.is_equal_pair(val1, v2)?;
+                }
+                (Value::Value(v1), val2) => {
+                    let v1 = self.get_value(v1);
+                    val = self.is_equal_pair(v1, val2)?;
                 }
-                _ => {}
+                (_, _) => {} // TODO XXX: are these cases meaningful?
```

Comment:
Yes, but the initial check (at the top) val1 == val2 will handle most of this.  For errors we might need to explicitly match since it is possible to create the "same" error as two different heap object.  For the other stuff I think it is very unlikely to have different heap objects for the same callable.  It may be possible (for instance compile same function twice) but even that case it may be ok to consider those different things...